File: LinkedList/ReverseLL.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};
struct Node* head; //global variable

//reverse a LL
void reverse(){
  Node *current, *prev, *next;
  prev = NULL;
  current = head;
  while(current != NULL){
    next = current->next;
    current->next = prev;
    prev = current;
    current = next;
  }
  head = prev;
}

//insert at the end
void append(int value){
  Node* newNode = new Node();
  newNode->data = value;
  newNode->next = NULL;

  Node* temp = head;
  while(temp->next != NULL){
    temp = temp->next;
  }
  temp->next = newNode;
  
}

int main(){
   
    head = NULL; //list is empty at this stage

     
    append(5);  
    append(15);  
    append(45);  
    append(35);  
    printList(); 

    reverse();
    printList(); 
    return 0 ;
}

File: LinkedList/ReverseLL.js
// add a method remove() to the linked list that deletes a node to the specified index.

class LinkedList {
    constructor(value) {
      this.head = {
        value: value,
        next: null,
        prev : null
      };
      this.tail = this.head;
      this.length = 1;
    }
    append(value) {
      const newNode = {
        value: value,
        next: null,
        prev : null
      }
      newNode.prev = this.tail;
      this.tail.next = newNode;
      this.tail = newNode;
      this.length++;
      return this;
    }
    prepend(value) {
      const newNode = {
        value: value,
        next: null,
        prev: null
      }
  
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
      this.length++;
      return this;
    }
    printList() {
      const array = [];
      let currentNode = this.head;
      while(currentNode !== null){
          array.push(currentNode.value)
          currentNode = currentNode.next
      }
      return array;
    }
    insert(index, value){
      //Check for proper parameters;
      if(index >= this.length) {
        console.log('yes')
        return this.append(value);
      }
      
      const newNode = {
        value: value,
        next: null,
        prev: null
      }
      const leader = this.traverseToIndex(index-1);
      const follower = leader.next;
      leader.next = newNode;
      newNode.prev = leader;
      newNode.next = follower;
      follower.prev = newNode;
      this.length++;
      return this.printList();
    }
    traverseToIndex(index) {
      //Check parameters
      let counter = 0;
      let currentNode = this.head;
      while(counter !== index){
        currentNode = currentNode.next;
        counter++;
      }
      return currentNode;
    }
    remove(index) {
      // Check Parameters      
      const leader = this.traverseToIndex(index-1);
      const unwantedNode = leader.next;
      const follower = unwantedNode.next;
      leader.next = follower;
      follower.prev = leader;
      this.length--;
      return this.printList();
    }
    //reverse function
    reverse(){
      let current = this.head;
      let prev = null;
      let next1;
      while(current!=null){
        next1 = current.next;
        current.next = prev;
        prev = current;
        current = next1;
      }
      this.head = prev;
      return this.printList();
    }
  
  }
  
  let myLinkedList = new LinkedList(10);
  myLinkedList.append(5);
  myLinkedList.append(16);
  myLinkedList.prepend(1);
  myLinkedList.insert(2, 99);
  myLinkedList.insert(4, 88);
  myLinkedList.reverse();
  
  
  
  